src/node_url.cc

```diff
@@ -407,29 +407,64 @@ int64_t ParseIPv4Number(const char* start, const char* end) {
   return strtoll(start, nullptr, R);
 }
 
+// https://url.spec.whatwg.org/#ipv4-number-parser
+bool IsIPv4NumberValid(const std::string_view input) {
+  if (input.empty()) {
+    return false;
+  }
+
+  // If a number starts with '0' it might be a number with base 8 or base
+  // 16. If not, checking if all characters are digits proves that it is a
+  // base 10 number.
+  if (input.size() >= 2 && input[0] == '0') {
+    if (input[1] == 'X' || input[1] == 'x') {
+      if (input.size() == 2) {
+        return true;
+      }
+
+      return input.find_first_not_of("0123456789abcdefABCDEF", 2) ==
+             std::string_view::npos;
+    }
+
+    return input.find_first_not_of("01234567", 1) == std::string_view::npos;
+  }
+
+  return std::all_of(input.begin(), input.end(), ::isdigit);
+}
+
 // https://url.spec.whatwg.org/#ends-in-a-number-checker
-bool EndsInANumber(const std::string& input) {
-  std::vector<std::string> parts = SplitString(input, '.', false);
+inline bool EndsInANumber(const std::string_view input) {
+  if (input.empty()) {
+    return false;
+  }
+
+  char delimiter = '.';
+  std::string_view::iterator pointer_start = input.begin();
+  std::string_view::iterator pointer_end = input.end();
 
-  if (parts.empty()) return false;
+  uint8_t parts_size = std::count(pointer_start, pointer_end, delimiter);
+  ++parts_size;
 
-  if (parts.back().empty()) {
-    if (parts.size() == 1) return false;
-    parts.pop_back();
+  if (input.back() == delimiter) {
+    --pointer_end;
+    --parts_size;
   }
 
-  const std::string& last = parts.back();
+  if (parts_size > 1) {
+    pointer_start +=
+        input.rfind(delimiter, pointer_end - pointer_start - 1) + 1;
```

Comment:
You're scanning the input string repeatedly in this function. Try to do it all in one forward scan, that's going to be both faster and easier to understand.

Reply:
I don't follow @bnoordhuis. The first iteration counts, and later this function finds the last `.` character. Counting and finding the last 2 dots (depending if the string ends with `.` or not), would be a lot slower (in my honest opinion)

Reply:
I'm not sure how could I do it (but I'm still thinking about it). But if I did these operations in an explicit loop (for/while with a counter, etc), we would lose all chance of the "inline" keyword being used by the compiler

Reply:
How is inline relevant here? Can you explain?

Reply:
I don't know, please forget what I said. I read somewhere else that compilers refuse the `inline suggestion` in functions that contain loops (for, while or do-while), but I don't know if it was an official reference or if this is true.

Reply:
And I think inline is relevant because it's an optimization that makes things faster... but I don't really know if it would apply in this case. In short, I need to study :sweat_smile:

Reply:
anyways I'll try to do what you're saying : )

Reply:
```c++
inline bool EndsInANumber(const std::string_view input) {
  if (input.empty()) {
    return false;
  }

  char delimiter = '.';
  auto last_index = input.size() - 1;
  if (input.back() == delimiter) {
    --last_index;
  }

  std::string_view last{};
  auto pos = input.find_last_of(delimiter, last_index);
  if (pos == std::string_view::npos) {
    last = input.substr(0, last_index);
  } else {
    last = input.substr(pos + 1, last_index - pos);
  }

  if(last.empty()) {
    return false;
  }

  if (std::all_of(last.begin(), last.end(), [](const char& c) {
        return IsASCIIDigit(c);
      })) {
    return true;
  }

  return IsIPv4NumberValid(last);
}
```
@bnoordhuis With less pointer shuffling + less loops. Looks better?

Reply:
should I switch to this implementation?

Reply:
can't decide between:

```c++
  if (std::all_of(last.begin(), last.end(), [](const char& c) {
        return IsASCIIDigit(c);
      })) {
    return true;
  }
```
and
```c++
  if (std::all_of(last.begin(),
                  last.end(),
                  static_cast<bool (*)(const char&)>(IsASCIIDigit))) {
    return true;
  }
```

Reply:
The following should work:

```
 if (std::all_of(last.begin(), last.end(), IsASCIIDigit) {
    return true;
  }
```

Reply:
It wasn't working on my Mac. I'll post the  compilation error here soon

Reply:
Ah, it's a macro, not a function.

Reply:
Which implememtation do you think we should go with? The one that is actually pushed or the last one I sent in this thread ?  (me: the last one I sent in this thread)

Reply:
It’s probably best to go with the first one with the lambda around `IsASCIIDigit()`.

I think the other one should also work but this one doesn’t rely on how `IsASCIIDigit` is implemented exactly and in terms of what they compile to they should give the exact same code.